pkg/statedb/reconciler/types.go

```diff
@@ -0,0 +1,207 @@
+	// the context error.
+	WaitForReconciliation(context.Context) error
+}
+
+type Config[Obj any] struct {
+	// FullReconcilationInterval is the amount of time to wait between full
+	// reconciliation rounds. A full reconciliation is Prune() of unexpected
+	// objects and Update() of all objects. With full reconciliation we're
+	// resilient towards outside changes. If FullReconcilationInterval is
+	// 0 then full reconciliation is disabled.
+	FullReconcilationInterval time.Duration
+
+	// RetryBackoffMinDuration is the minimum amount of time to wait before
+	// retrying a failed Update() or Delete() operation on an object.
+	// The retry wait time for an object will increase exponentially on
+	// subsequent failures until RetryBackoffMaxDuration is reached.
+	RetryBackoffMinDuration time.Duration
+
+	// RetryBackoffMaxDuration is the maximum amount of time to wait before
+	// retrying.
+	RetryBackoffMaxDuration time.Duration
+
+	// IncrementalBatchSize is the maximum number objects to reconcile during
+	// incremental reconciliation before updating status and refreshing the
+	// statedb snapshot. This should be tuned based on the cost of each operation
+	// and the rate of expected changes so that health and per-object status
+	// updates are not delayed too much. If in doubt, use a value between 100-1000.
+	IncrementalBatchSize int
+
+	// GetObjectStatus returns the reconciliation status for the object.
+	GetObjectStatus func(Obj) Status
+
+	// WithObjectStatus returns a COPY of the object with the status set to
+	// the given value.
+	WithObjectStatus func(Obj, Status) Obj
+
+	// StatusIndex is the index for looking of objects by StatusKind. The index
+	// is created with NewStatusIndex.
+	StatusIndex statedb.Index[Obj, StatusKind]
+}
+
+func (cfg Config[Obj]) validate() error {
+	if cfg.GetObjectStatus == nil {
+		return fmt.Errorf("%T.GetObjectStatus cannot be nil", cfg)
+	}
+	if cfg.WithObjectStatus == nil {
+		return fmt.Errorf("%T.WithObjectStatus cannot be nil", cfg)
+	}
+	if cfg.StatusIndex.Name == "" {
+		return fmt.Errorf("%T.StatusIndex needs to be defined", cfg)
+	}
+	if cfg.IncrementalBatchSize <= 0 {
+		return fmt.Errorf("%T.IncrementalBatchSize needs to be >0", cfg)
+	}
+	if cfg.FullReconcilationInterval < 0 {
+		return fmt.Errorf("%T.FullReconcilationInterval must be >=0", cfg)
+	}
+	if cfg.RetryBackoffMaxDuration <= 0 {
+		return fmt.Errorf("%T.RetryBackoffMaxDuration must be >0", cfg)
+	}
+	if cfg.RetryBackoffMinDuration <= 0 {
+		return fmt.Errorf("%T.RetryBackoffMinDuration must be >0", cfg)
+	}
+	return nil
+}
+
+// Operations defines how to reconcile an object.
+//
+// Each operation is given a context that limits the lifetime of the operation
+// and a ReadTxn to allow for the option of looking up realized state from another
+// statedb table as an optimization (main use-case is reconciling routes against
+// Table[Route] to avoid a syscall per route).
+type Operations[Obj any] interface {
```

Comment:
For all of the operations, if the given `context.Context` is canceled, what should be the behavior for the implementations? Should the revert or should they simply "stop" that operation?

Reply:
I've honestly not thought of its semantics much since all the reconcilers thus far have done just quick uninterruptible syscalls. The context is really just telling that hey we're shutting down and please try to stop. So it's probably mostly applicable for long-running cancellable operations like doing a REST API call or something. I think "simply stop" is likely most we can ask for here.